rfc-128-continuous-deployment.md

```diff
@@ -0,0 +1,237 @@
+# Continuous Deployment
+
+## Summary
+
+In [RFC 70](https://github.com/alphagov/govuk-rfcs/blob/master/rfc-070-path-towards-continuous-deployment-cd.md#proposal) we committed to fully automatic deployments. The following provides a framework for how we will actually achieve this for all of our existing apps.
+
+Decision: in order to safely enable automatic deployments for each app, we will specify minimum criteria that must first be met. See the criteria described [below](#safety-checks).
+
+An audit of our apps against these criteria can be found in [Appendix A: Compatibility](rfc-128/compatibility.md).
+
+## Problem
+
+Our current approach to shipping software has multiple issues with it:
+
+### 1. We have a constant backlog of deployments
+
+Merging a PR for an application repo [results in a new "release" tag](https://github.com/alphagov/govuk-jenkinslib/blob/30e9392d31b98e5c251ea471d793761748eea1e8/vars/govuk.groovy#L123). We have a [Release](https://github.com/alphagov/release) app to monitor these. Every new release tag [is automatically deployed to our Integration environment](https://github.com/alphagov/govuk-jenkinslib/blob/30e9392d31b98e5c251ea471d793761748eea1e8/vars/govuk.groovy#L134).
+
+This is where the automation ends. In order for a release (or series of releases) to be deployed to the GOV.UK Production environment, a developer with [production access](https://docs.publishing.service.gov.uk/manual/rules-for-getting-production-access.html) must:
+
+- Manually check the release is "OK" in Integration.
+- Manually deploy the release to Staging.
+- Manually check the release is "OK" in Staging.
+- Manually deploy the release to Production.
+
+Having these tedious, manual steps makes it easy to *not* deploy a change. This is especially true for new joiners, who can only merge PRs and cannot deploy them beyond Integration.
+
+> Full stats and supporting code can be found [here](https://docs.google.com/spreadsheets/d/1Hw1qE-ozqb7n97-fsWkH8Ddgz-7Jyr2G8z8lHXJZPhE/edit#gid=0). We expect these will improve slightly over the coming months, due to the reduction in changes generated by Dependabot [[1](https://github.com/alphagov/govuk-rfcs/blob/master/rfc-126-custom-configuration-for-dependabot.md)].
+
+![Graph of deploy lag (in days)](rfc-128/chart.svg)
+
+A backlog of deployments also makes diagnostics harder:
+
+- ⚠ We can mistakenly debug an issue against the undeployed version of the code.
+- ⚠ We may struggle to isolate a breaking change, if many were deployed together.
+
+### 2. We freely merge PRs with breaking changes
+
+We tend to merge PRs and only later worry if they break the app. This can cause the codebase to become undeployable, and interrupt other developers testing changes on Integration.
```

Comment:
I'm finding problem statement is bit vague and difficult to understand.

What I think what you are saying here is: Bugs/issues are merged in to the deployment branch (master/main). These bugs are discovered after commits have been merged, but before they have been deployed to production. This prevents future deployments until bug fixes have also been merged into the deployment branch.

Do we know how we discover bugs after they've been merged? ("only later worry if they break the app"). I assume this when we manually test and run the code the integration environment? (Which gives reason to making sure we don't need to manually test and we write adequate integration tests)

 How does this "interrupt other developers testing changes on Integration"? I thought we are still able to deploy any branch integration.

Reply:
```suggestion
### 2. Discovering issues/bugs after merging to the deployment branch

Sometimes we merge commits with issue or bugs into our deployable branch (main/master). These issues are sometimes only discovered after the commits has been merged, but before they have been deployed to production. This prevents any future deployments to production until a fix has been merged into the deployment branch.  This can cause delays in deployment of other code changes for example new features or other bug fixes. 
```

Wasn't too sure what was meant by "interrupt other developers testing changes on Integration"?

Reply:
Thanks for the suggestion - very helpful 💯. I think it's quite different from what I'm trying to say, though. I'll try and talk around it, which might help: we will always find bugs after we merge changes; the problem is that we currently put too little effort into finding them _before_ we merge i.e. we click "merge" without thinking (much).

This is kind of separate to the problem of adequate testing, which we should try to be objective about if possible. What I'm trying to get across here is the other side of testing: the human intelligence we _can't_ be objective about, where we think deeply about the change we're making and any subtle, hard to test ways it might break something. At the moment, I think we freely merge our PRs without applying much intelligence at the point we click the button.

It might help to point out the parts of the main text that address this:

> The person who merges a PR must be confident that it is safe to do so.

> We are ultimately responsible for our changes, and will need to adapt to this new "CD" way of working by ...

I think we've seen already how using CD has prompted us to be more cautious i.e. apply our intelligence before we click merge, instead of after. So, conceptually the current text feels OK to me, but perhaps the wording could do a better job of explaining the problem (like you say in your comment).

First thing to check: does what I've said make sense?

Reply:
Thanks Ben! From what you’ve said the problem is that : 

> we currently put too little effort into finding them [issues] before we merge i.e. we click "merge" without thinking (much).

and that leads to:  

> Causes the codebase to be un-deployable and interrupt developers testing changes in Integration.

  the latter which I’m understanding to be ultimately one of the problems we are trying to solve with CD.  

What could make this section clearer - is describing what actually causes a codebase to be undeployable? and why does this interrupt developer testing on Integration?   I think my suggestion was trying to explain how the codebase becomes "undeployable" at a higher level. I.e. it only becomes undeployable because we become aware of an issue after merging but before production. 

I'd like to argue that we should make this problem statement higher level because:

- It makes it easier to understand the problem we are solving i.e. its hard to see why "Freely merging PRs" is a problem vs having a "undeployable" codebase.
- Makes it easier to reason a solution - if a un-deployable codebase is because we've merged a bug, but not yet gone live - what do currently do to catch these issues after merging? and can we move them to before we merge?
- "Freely merging PRs" isn't the sole factor to the underlying problem. Inadequate automated testing and manually testing after merging are probably also contributing factors. 

Let me know if I’ve got the wrong end of the stick!

Reply:
Hmm...this is a tricky one. I was struggling to understand the request you're making. I think it's because there are a couple of ways we can cut the problems we have in these sections:

1. In terms of the types of testing we do (including manual testing).
2. In terms of our attitude towards (a) merging and (b) deploying.

I think you're focussing on (1) and I'm focussing on (2). Both cover the same things, really:

- Our attitude towards merging affects how much testing we do before this point. This can be manual or automatic, and they are both motivated by how we perceive the risk of clicking "merge".

- Our attitude towards deploying affects the post-merge checks we do. Some tests are easier to run at this point, but before we deploy to production. These can be manual or automatic as well.

If we cut with (1), then these sections would be something like "we don't know if we have enough tests" and "our manual tests are too inconsistent to codify". However, I find the people angle easier to write about.

How about this?

### We seldom verify our changes before merging

We tend to merge PRs and only later worry if they break the app. This will range from a feature that no longer works as expected, to the app not running at all. Once a breaking change is merged, we cannot deploy any newer changes without including the broken one. Anyone trying to use the app for testing may also be unable to do so.

Although we have ways to verify a change before we merge it - branch deployments and more internal repo tests - we often rely on our manual deployment process as the "great opportunity" to find bugs. This is unlikely to be as rigorous: the time gap between writing code and deploying the change means we lose context of how to test it.

[append rest of section]

Reply:
Yup - thats fine!

Nit: Maybe have "We sometimes verify our changes after merging” as the title - i.e. drop "seldom". It comes across as a bit of negative generalisation "all developers rarely check their changes before merging".

Reply:
Thanks @theseanything! I'm wondering what to do now. I could make another PR to amend the section. An alternative would be to leave this comment unresolved, so it can still be found easily. I don't mind.

With "seldom" vs. "sometimes", it is the negative I'm aiming for. Perhaps a more accurate phrase would be "We avoid verifying our changes before we merge". I think that's a better match for what follows.

What do you think?